### ISSUE TITLE
Gradle 7 requires duplicatesStrategy for "fake" duplicates

### ISSUE THREAD

**Issue by lkoe** [gradle/gradle#17236]
Gradle 7 rightfully requires a `duplicatesStrategy` for handling duplicate files.

However, the respective error can also be produced by adding the same directory to a source set multiple times.
This is a situation often encountered when using multiple third-party plugins, which add the same directories to a source set.

This situation is mostly out of control of the user, yet requires to specify a handling strategy - which in this case seems ultimately unnecessary, since no real file duplicates (duplicate files in multiple directories) are involved.

`SourceDirectorySet#getSourceDirectories()` might be a potential point, where duplicate source directories could be considered and filtered.

### Expected Behavior
Gradle should allow innocently adding the same source directory multiple times, without requiring to set `duplicatesStrategy` for common tasks (`processResources`, `sourcesJar`)

### Current Behavior
Gradle throws the duplicatesStrategy error, when the same source directory is added multiple times.

### Context
We are using complex builds with a variety of third-party plugins with little to no control over them. The problem forced us to update our build scripts/custom plugins, even though there weren't any actual duplicates involved.

### Steps to Reproduce 
Build the attached sample project.
[g7-duplicate-sources.zip](https://github.com/gradle/gradle/files/6514424/g7-duplicate-sources.zip)

### Your Environment
Gradle 7.0/7.0.2, Windows 10
Build scan URL:

**Comment by lackovic** [comment]
I am also facing this issue while upgrading from Gradle `6.6.1` to `7.0.2`, with no real file duplicates involved:
```
Task :compileE2eTestJava
[Step 1/1] :compileE2eTestJava
[:compileE2eTestJava] FAILURE: Build failed with an exception.
[:compileE2eTestJava] 
[:compileE2eTestJava] * What went wrong:
[:compileE2eTestJava] Execution failed for task ':processE2eTestResources'.
[Step 1/1] > Entry application.properties is a duplicate but no duplicate handling strategy has been set. Please refer to https://docs.gradle.org/7.0.2/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:duplicatesStrategy for details.
[Step 1/1] > Task :processE2eTestResources FAILED
[Step 1/1] * Try:
[Step 1/1] Run with --info or --debug option to get more log output. Run with --scan to get full insights.
[Step 1/1] :processE2eTestResources
[:processE2eTestResources] * Exception is:
[:processE2eTestResources] org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':processE2eTestResources'.
```

The note to _refer to https://docs.gradle.org/7.0.2/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:duplicatesStrategy for details_ is not helping since the task `processE2eTestResources` is defined by Gradle, not by me, so it is unclear to what exactly I need to add this property to in order to fix this issue.

**Comment by donat** [comment]
This is a manifestation of a missing internal Gradle feature. Gradle does not do any deduplication in this use case. And even if such a feature would exist, this configuration would cause many issues. Tasks cannot be up-to-date, caching won't work, etc.

It is the plugin's responsibility to define unique input and output folders. If there's such an overlap, the plugins should also enable the clients to configure those folders.

**Comment by lkoe** [comment]
Agreed that this kind of overlap can be troublesome for other Gradle features, like you mentioned.

But the issue at hand is that Gradle is flagging duplicates, when in fact there aren't any. I still think this should be adressed.

**Comment by dimaKudr** [comment]
@donat from the Gradle user perspective it looks like a defect introduced in Gradle 7. It would be very nice for the Gradle team to mention it in the Upgrade Guide here https://docs.gradle.org/current/userguide/upgrading_version_6.html. Its current section about Duplication Strategy is kind of misleading. It says "encounters a duplicate entry", and when I read that, I think "but I don't have any duplicates, so it is not relevant to me". Hope you see my point.
And fix for the problem was suggested here https://discuss.gradle.org/t/gradle-7-fail-for-duplicates-in-copy-specs-has-no-duplicates-in-project/39834/8

**Comment by unoexperto** [comment]
Folks, what is the solution for this problem? I'm honestly baffled. I'm using 7.1. I followed advice from official upgrade documentation and added this

```
tasks {
    withType<Copy> {
        duplicatesStrategy = DuplicatesStrategy.EXCLUDE
    }
}
```

But it doesn't seem to help, as if this setting is not inherited by other tasks. So that `distTar` or `assemble` fail for me with the same error unless I set `duplicatesStrategy` for each of these tasks.

**Comment by mbwhite** [comment]
In case it helps. I believe I hit exactly this problem, and the solution that seems to work for me is 

```
task sourcesJar(type: Jar) {
    duplicatesStrategy = 'include'
    classifier = 'sources'
    from sourceSets.main.allSource
}

java {
    withJavadocJar()
    withSourcesJar()
}
```

Order of these is important! `task sourcesJar` must come first, and you must also have `withSourcesJar()`

**Comment by ylemoigne** [comment]
For anyone coming here and using kotlin multiplatform, the plugin trigger the issue if `withJava()` is used : https://youtrack.jetbrains.com/issue/KT-46978

The official gradle response to this issue is very disapointing. It's ok to don't fix it as plugin should not do what they are doing.
But showing some level of empathy for your users would be greatly appreciated.

Actually we have a message wich talk about a duplicated resource that is not duplicated. It's missleading.
Their is nothing in documentation/upgrade-guide that can help.
Users can spend hours searching in documentation/web before finding this issue (or the forum post mentionned)
If this issue didn't exist or if user don't find it, they don't understand what happen, they feel stupid for their not being able to make their project work. Or they curse gradle for being "too complicated" and "not working".

This not very pleasant for everyone.

Once this issue is found, it's time to remove plugins, comment any build part depending on it, and re-add/enable them one by one, maybe in different combination to search culprit(s). On complex build, it's cumbersome.

What can be done ?
At level 1, mention in upgrade guide that some plugin which was incorrectly using copy might break the build.
At level 2, mention in the error message that it might be caused by plugin bugs.
At level 3, detect directory added multiple times, produce a message adapted to this situation
At level 4, detect directory added multiple times, and say which plugin(s)/call-site is adding it

I understand that it may be not trivial to handle.
But saying "not our problem : good bye" when issue is not directly under user control nor visibility of user and triggered by and update of gradle, letting user suffering from it, it feel like gradle don't care and it's just sad...

**Comment by robvadai** [comment]
@ylemoigne I completely agree.

Gradle guys, the issue is if Gradle gets too overwhelming to use it'll loose popularity. I don't want that, I love to use Gradle for many years now but this is a warning sign. Please make sure releases are user friendly.

I had genuine duplicate files in my repo (`log4j.properties` in different test directories) and my solution was:

```groovy
allprojects {
  apply plugin: 'java'

  tasks.withType(Copy).all {
    duplicatesStrategy 'exclude'
  }
...
```

**Comment by lackovic** [comment]
I had limited time to troubleshoot my issue mentioned above so eventually I had to rollback to version 6.

A possible workaround (not tested by me yet), suggested by [this comment](https://discuss.gradle.org/t/gradle-7-fail-for-duplicates-in-copy-specs-has-no-duplicates-in-project/39834/8), is to add the following to the `build.gradle`:

```js
rootProject.tasks.named("processE2eTestResources") {
    duplicatesStrategy = 'include'
}
```

**Comment by calvertdw** [comment]
Kotlin type safe syntax that worked for me:
```
<project>.tasks.named("processResources", Copy::class.java) {
   duplicatesStrategy = DuplicatesStrategy.INCLUDE
}
```

**Comment by GregMarshBYKA** [comment]
@JiangHongTiao - I contacted jetbrains support and they pointed back to gradle so I see no point trying to come up with a work around with gradle for it does not make sense to put in extra code throughout a project just to fix a build issue. 

Hope gradle/jetbrains figures out what is actually going wrong with 7 to 7.2 ....

**Comment by jhyry-gcpud** [comment]
Still having this issue in 7.2.

**Comment by gwitsch** [comment]
Just had the same issue while upgrading from Gradle 6.7 to 7.3.
It has been caused by my sourceSets definition
`sourceSets {
	main {
		java {
			srcDirs = ['src']
		}
		resources {
			srcDirs = ['src']
			includes = ['**/*.xsd']
		}
	}
}`

I fixed it in my scripts by setting the duplications strategy for the sourcesJar in my custom plugin:
`project.getTasks().getByName("sourcesJar").setProperty("duplicatesStrategy", DuplicatesStrategy.EXCLUDE);`

Maybe this is helpful for others, that come across this issue too.

**Comment by ryantheleach** [comment]
I'm getting this when upgrading from a previous gradle version.

```
processResources {
    from(sourceSets.main.resources.srcDirs) {
        filter ReplaceTokens, tokens: [version: version]
    }
}

```

seemed to be a popular example for filtering something bundled in the resources folder before it was packaged.

However this error indicates it's creating a duplicate?

**Comment by leonardBang** [comment]
Fortunately, I found this issue here, otherwise I will be stunned by gradle 7.

**Comment by chjan** [comment]
After upgrading from some 6.x to 7.3.3, I get this error report while building a pure java module. It is beyond me to understand what is the concrete cause, as there is no pointing to line number, nor to the dependency causing this trouble. After a day of trials in vain, I would like to announce a strong wish to be presented for better error messages. What a timethief for something that seems to be most likely on the fringe of the build process, regarding importance.

I ask @mlopatkin or @donat to be so kind to make a suggestion on how to handle this. Thanks

**Comment by GregMarshBYKA** [comment]
Just wondering if this is being worked on? I ask because I can't deploy a training program that involved using Gradle as part of the development tools. As per my previous note I did revert to a previous version but given the seriousness of this issue I figured a fix would be issued in short order. 
Am I missing something or does this issue not prevent people from using Gradle 7?  

If it is not going to be fixed or is considered a low priority can this be stated so I and people like me can make a decision as to which direction to go? I have to make a decision soon and had hoped for a fix.

**Comment by swpalmer** [comment]
@GregMarshBYKA You are missing something.  Plenty of people are using Gradle 7.  This issue, while annoying, can be worked around in the interim while we are waiting for a proper fix.  You have to set the duplicatesStrategy in one of various ways (plenty of examples above) and you can continue on with Gradle 7.

**Comment by GregMarshBYKA** [comment]
@swpalmer 
I don't think I am missing anything. As per my original post, I will be starting several training programs for coders and we had intended to use the initial demo which currently does not work out of the box.
As to the solutions, these are workarounds and I have not checked all of them - a few of them required a lot of redundant work.
I was hoping to get clarity as to when the fix is scheduled. 
As to the latest fix mentioned I will try that and the following might be some value for those working on a solution.
https://docs.gradle.org/current/javadoc/org/gradle/api/file/DuplicatesStrategy.html
Thank you

**Comment by rivancic** [comment]
I have bumped on this new behavior as well. I had project with several subprojects. And many of them had the same file in the resource directory (.giitkeep) as there was no files in the directory but we still wanted to keep it in the repository. Now during the generation of jar artifact, this exception has been thrown. 
In my case simple solution was to remove those .gitkeep files so there were no duplicate files anymore in the final artifact.

In case there would really be some files that I would require in final artifact I would need to find a way of how to exclude them, or just follow what @lackovic suggested https://github.com/gradle/gradle/issues/17236#issuecomment-894768083

**Comment by tonicsoft** [comment]
In my case the issue was not the duplicate handling. Fair enough, silently overwriting duplicate files was bad. In my case it has revealed a particular fact about our build that I cannot explain, and really don't know where to start debugging. The best I could achieve while debugging with Intellij was that it might have something to do with the fact we are using the google protobuf plugin.

The Jar task was trying to copy 3 identical sets of files and I have no idea why. So the new duplicateStrategy has kind of revealed some other issue here, but it was annoying nonetheless.

```kotlin
// package the compiled application, along with all dependencies, into a single runnable jar
val fatJarTask = tasks.register<Jar>("app") {
    archiveClassifier.set("app")

    from(sourceSets.main.get().output)

    dependsOn(configurations.runtimeClasspath)
    from({
        logger.error("this message is printed 3 times when running the 'app' task and it's not clear why")
        configurations.runtimeClasspath.get().filter { it.name.endsWith("jar") }.map { zipTree(it) }
    })

    // https://github.com/gradle/gradle/issues/17236
    duplicatesStrategy = DuplicatesStrategy.INCLUDE

    manifest {
        attributes["Main-Class"] = "MainKt"
    }
}
```

**Comment by shalom938** [comment]
I just hit this issue too with gradle 7.3.3. I'm using the jvm-test-suite plugin, i have a 'test' suite and 'integrationTest' suite.
as soon as i added log4j2.xml to src/integrationTest/resources i got this error:

> Execution failed for task ':messenger:processIntegrationTestResources'.
>  Entry log4j2-test.xml is a duplicate but no duplicate handling strategy has been set. Please refer to https://docs.gradle.org/7.3.3/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:duplicatesStrategy for details.
> 

for sure there are no duplicates , this was my workaround:
processIntegrationTestResources{
    duplicatesStrategy = DuplicatesStrategy.WARN
}

**Comment by kmayura1** [comment]
Thanks for all previous comments,  this worked for me when we have a task already defined.
```
processResources() {
    duplicatesStrategy = 'EXCLUDE'
}
```

**Comment by peter-lawrey** [comment]
I added the workaround previously mentioned by @robvadai and it worked. 
This isn't my project, https://github.com/Kotlin/kotlinx.serialization
I am not familiar with Gradle and I just want it to compile.

allprojects {
    apply plugin: 'java'

    tasks.withType(Copy).all {
        duplicatesStrategy 'exclude'
    }
}

**Comment by NathanielRN** [comment]
For those looking for a kotlin solution (i.e. how to write this in a `build.gradle.kts` file) [this worked for us](https://github.com/aws/aws-xray-sdk-java/blob/2e22dcb484fe904f27468d080ba4783610802bd2/aws-xray-recorder-sdk-benchmark/build.gradle.kts#L31-L36):

```kotlin
tasks.jmhJar {
    duplicatesStrategy = DuplicatesStrategy.EXCLUDE
}
```

**Comment by johnlabarge** [comment]
none of the above solutions seem to work and I always get a build failure with this Entry BOOT-INF/lib/reactive-streams-1.0.3.jar is a duplicate but no duplicate handling strategy has been set. Please refer to https://docs.gradle.org/7.4.2/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:duplicatesStrategy for details.

rolling back to gradle 6, again.

**Comment by sdykae** [comment]
help: after https://stackoverflow.com/questions/67265308/gradle-entry-classpath-is-a-duplicate-but-no-duplicate-handling-strategy-has-b
with: `Entry WEB-INF/lib/ObjectFactory.class is a duplicate but no duplicate handling strategy has been set.`

Realized that in my case I needed to do this in the type Jar instead Copy, thank you all  
```groovy
tasks.withType(Jar).all {
    duplicatesStrategy 'exclude'
  }
```

**Comment by uap-universe** [comment]
I don't know the current status and I did not invest too much time researching, so I just share my observation in case in helps someone. 

I have a Kotlin project with the build file below. With gradle 7.3 I encountered an "Entry classpath.index is a duplicate but no duplicate handling strategy has been set" error. Since it clearly cannot be something that I configured in the build file, I refused to add any config. The build *has to work* just like that. So I updated to gradle 7.6 and hoped for the best and yep - the error was gone. And don't know why, I don't know what caused it in the first place, and I honestly don't care.

Since this thread is full of "add this to your config" suggestions I just wanted to share that it is clearly not always the "fix" to this issue to take the blame and change the build config. There must have been a (or at least one) bug in the Gradle software that has been fixed.

```
import org.jetbrains.kotlin.gradle.tasks.KotlinCompile

plugins {
    kotlin("jvm") version "1.7.21"
    war
}
group = "de.uapcore"
version = "1.0-SNAPSHOT"

repositories {
    mavenCentral()
}

tasks.withType<KotlinCompile>().configureEach {
    kotlinOptions.jvmTarget = "11"
}

tasks.war {
    archiveFileName.set("lightpit.war")
    from("src/main/resources")
}

kotlin {
    sourceSets {
        val main by getting {
            dependencies {
                // ... usual suff ...
            }
        }
    }
}

```

**Comment by docent** [comment]
> I don't know the current status and I did not invest too much time researching, so I just share my observation in case it helps someone.
> 
> I have a Kotlin project with the build file below. With gradle 7.3 I encountered an "Entry classpath.index is a duplicate but no duplicate handling strategy has been set" error. Since it clearly cannot be something that I configured in the build file, I refused to add any config. The build _has to work_ just like that. So I updated to gradle 7.6 and hoped for the best but no, the error was still there.
> 
> **UPDATE - I did some research now...**
> 
> Since this thread is full of "add this to your config" suggestions I wanted to find a solution that does not require the user to add mysterious config to the build file. There _has to be_ a way, otherwise the build system would not work with a default config, which would be really really bad.
> 
> Apparently in my case the root cause is that most recent versions of IntelliJ are generating some classpath.index file that Gradle does not expect and does not know how to handle. And since this file is generated in every source file location (i.e. at least one in code directory and one in resource directory), it is a genuine duplicate.
> 
> See https://youtrack.jetbrains.com/issue/IDEA-305759/Gradle-cannot-handle-classpath.index-duplicates for more info.
> 
> To be very clear: I still believe that Gradle _should_ have proper defaults that do not cause such builds to fail. But I also would admit that currently Gradle works as designed. It's just poor design imho...
> 
> ```
> import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
> 
> plugins {
>     kotlin("jvm") version "1.7.21"
>     war
> }
> group = "de.uapcore"
> version = "1.0-SNAPSHOT"
> 
> repositories {
>     mavenCentral()
> }
> 
> tasks.withType<KotlinCompile>().configureEach {
>     kotlinOptions.jvmTarget = "11"
> }
> 
> tasks.war {
>     archiveFileName.set("lightpit.war")
>     from("src/main/resources")
> }
> 
> kotlin {
>     sourceSets {
>         val main by getting {
>             dependencies {
>                 // ... usual suff ...
>             }
>         }
>     }
> }
> ```

Thanks for pointing this out, it turned out to be the reason for me.
However I think that Gradle's duplicate handling strategy is fine, otherwise you would have silent duplicates. Failing fast in this case is good. Maybe the message could point to the files.

**Comment by caoccao** [comment]
I met the similar issue. The solution is quite simple: Execute `rm -rf <project>/build` to remove the `build` folder.

**Comment by docent** [comment]
@uap-universe In my opinion the default strategy should be to fail because I believe in "fail fast" approach. Avoid unexpected issues.

**Comment by uap-universe** [comment]
I did not try the FAIL strategy, yet. If you get a similar message as in WARN pointing to the files, I could live with that. But at least _some_ default would already make me happy.

**Comment by dal23** [comment]
Yes, same here. I was thinking classpath.index is a variable or configuration parameter. I didn't think about a file until I read this issue. I think
1. the message should be more clear. the error is obviously an error, so this is fine for me. probably better than silently ignoring
2. it should not be necessary to fix this by the developer. intellij did a change and this brakes a very widely used usecase. so I clearly think intellij should stop causing this issue.
3. best workaround for me is: "adding -Didea.classpath.index.enabled=false into [Help | Edit Custom VM Options](https://intellij-support.jetbrains.com/hc/en-us/articles/206544869) and Settings (Preferences on macOS) | Build, Execution, Deployment | Compiler | User-local build process VM options (overrides Shared options)." (https://youtrack.jetbrains.com/issue/IDEA-305759/Gradle-cannot-handle-classpath.index-duplicates#focus=Comments-27-6626901.0-0)

**Comment by GregMarshBYKA** [comment]
This thread contains a number of possible solutions that I assume the Gradle Team can judge as to which is the best option. I am just wondering what needs to resolve this issue which for myself I commented on back in Oct 2021. Back then I went back to a version of Gradle that worked "https://github.com/gradle/gradle/issues/17236#issuecomment-938892546" but it was not really a viable option. 
I hope that the consensus is to fix the issue versus providing "workarounds" that always get us into trouble. 
Either way, Happy New Year to all and I hope that this issue is finally laid to rest soon.

**Comment by uap-universe** [comment]
@GregMarshBYKA The least common denominator seems to be to identify the failing task (which might not be in your build file, yet! So that's the "hard" part - at least for beginners) and then explicitly specify the duplicate strategy to be WARN (some also go for EXCLUDE right away, but I'd prefer WARN).


Maybe @big-guy can shed some light on what happened since Nov 2021 what we can expect in a future version.

In my particular case, for example, I _removed_ the entire default JAR task. The funny thing about this is, that I have to explicitly write code, to disable something that breaks the build :-D 

### before
```Kotlin
tasks.war {
	archiveFileName.set("lightpit.war")
	from("src/main/resources")
}
```

### after

```Kotlin
// Only configuring the WAR task I want in my build is not enough, because
// there is a default JAR task, which also throws the error when there is no
// DuplicatesStrategy set. Therefore, I either need to ALSO configure the JAR
// task or disable it entirely (which is what I do - because I don't need it).
tasks.jar {
	enabled = false
}
 
tasks.war {
	duplicatesStrategy = DuplicatesStrategy.WARN
	archiveFileName.set("lightpit.war")
	from("src/main/resources")
}
```

**Comment by GregMarshBYKA** [comment]
> [uap-universe](/uap-universe)

Thank you, Mike, for your comments, for me the fundamental problem is that we are being asked to add a solution for something that does not exist and since this was not a problem in earlier versions I don't understand the justification for allowing such a workaround to exist. 
My ask was based on the simple fact that the introductory example used to build an app that used Gradle failed - IE:, the equivalent of C's "Hello World"  not working. 
I am happy to use a Gradle-approved solution, even if it is a forced workaround to solve the issue but maybe that is better as a flag that can be turned off and on.
Normally I would not bother and just move on to something else but there is something simply annoying about this bug in Gradle that keeps flagging duplicates that don't exist. I am assuming that the result has nothing to do with duplicates found (no matter how bizarrely they are created.) but rather code that fails and triggers this event.
As to letting it default to a warning that is a good solution simply because we can add code when duplicates actually exist and we can ignore adding code when we are starting from scratch or know that duplicates do not exist.
Does the Gradle team respond to these threads or is it simply user driven?
Cheers

**Comment by uap-universe** [comment]
> Thank you, Mike, for your comments, for me the fundamental problem is that we are being asked to add a solution for something that does not exist and since this was not a problem in earlier versions I don't understand the justification for allowing such a workaround to exist.

Yes, that was also exactly my point in a previous comment. I think we completely agree in all points. Now it's up to gradle devs to hopefully come up with a solution soon.

**Comment by GregMarshBYKA** [comment]
> > Thank you, Mike, for your comments, for me the fundamental problem is that we are being asked to add a solution for something that does not exist and since this was not a problem in earlier versions I don't understand the justification for allowing such a workaround to exist.
> 
> Yes, that was also exactly my point in a previous comment. I think we completely agree in all points. Now it's up to gradle devs to hopefully come up with a solution soon.

It has been at least a year so, yes, I hope so too for I would like to be able to use Gradle at some point for new development.

**Comment by GregMarshBYKA** [comment]
I really appreciate everyone's comments for it shows that this is clearly an issue and not something that is born out of a failure to understand how Gradle works.
That said, can anyone suggest an effective path to engage Gradle regarding this fundamental failure? Not knowing the inner workings of Gradle I do not want to assume that the solution is only dependent on resources being assigned to fix a bug. If it is complex and can't be fixed right now may we know why? 
Like many support threads, it is very frustrating to watch users fix or attempt to fix things that should be dealt with by those in the know. 
Fingers crossed.

**Comment by uap-universe** [comment]
@big-guy can you please confirm, that you are still working on this? @jvandort can you please have a look and maybe reassign this issue if @big-guy is currently busy doing something else? Thank you so much :-)

**Comment by fsparv** [comment]
If I understand the (quickly scanned) discussion above this indicates that some form of copy or similar task is writing a file that already exists. If that's true this has got to be the worst error message ever... "you have a duplicate Named X. Too bad, you figure it out" is about what it says. The great thing is that it's often the same thing being copied twice so no matter how much you grep or search for a duplicate, it's not there. Duplicate of what? where is the destination that seems to contain a duplicate? If gradle wants to ensure that one never accidentally overwrites via copy, then it needs to track all the things that copy files. (src/main/resources/foo.txt --> build/resourceTmp/foo.txt etc) Then when it tries to copy src/test/resources/foo.txt --> build/resourceTmp/foo.txt it can complain like this:

````
src/test/resources/foo.txt attempting to overwrite /build/resourceTmp/foo.txt (previously copied from src/main/resources/foo.txt) 
````

For even better clarity track the current task and modifed times so that one could understand

````
src/main/resources/foo.txt (task:mungeResources) attempting to overwrite /build/resourceTmp/foo.txt (previously copied from src/main/resources/foo.txt (task:copyResources) 
````
And possibly NOT fail if the mod time hadn't changed.

Wait, that's too much memory you say... possibly which is why the biggest problem here is this should be an optional feature for folks who are having issues with things getting mysteriously overwritten. Maybe something like

````
allProjects {
  strictCopy true
}
````

**Comment by notaLonelyDay** [comment]
Problem: 
`org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':mycoolmodule:processResources'.`
`Caused by: org.gradle.api.InvalidUserCodeException: Entry mycoolresource is a duplicate but no duplicate handling strategy has been set. Please refer to https://docs.gradle.org/7.5/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:duplicatesStrategy for details.`

Solution:
`processResources.duplicatesStrategy = DuplicatesStrategy.EXCLUDE`

**Comment by GregMarshBYKA** [comment]
> Problem: `org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':mycoolmodule:processResources'.`
> 
> `Caused by: org.gradle.api.InvalidUserCodeException: Entry mycoolresource is a duplicate but no duplicate handling strategy has been set. Please refer to https://docs.gradle.org/7.5/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:duplicatesStrategy for details.`
> 
> Solution: `processResources.duplicatesStrategy = DuplicatesStrategy.EXCLUDE`

Thank you but we are discussing a bug in Gradle whereby an error is thrown when a duplicate does not exist.

**Comment by GregMarshBYKA** [comment]
> If I understand the (quickly scanned) discussion above this indicates that some form of copy or similar task is writing a file that already exists. If that's true this has got to be the worst error message ever... "you have a duplicate Named X. Too bad, you figure it out" is about what it says. The great thing is that it's often the same thing being copied twice so no matter how much you grep or search for a duplicate, it's not there. Duplicate of what? where is the destination that seems to contain a duplicate? If gradle wants to ensure that one never accidentally overwrites via copy, then it needs to track all the things that copy files. (src/main/resources/foo.txt --> build/resourceTmp/foo.txt etc) Then when it tries to copy src/test/resources/foo.txt --> build/resourceTmp/foo.txt it can complain like this:
> 
> ```
> src/test/resources/foo.txt attempting to overwrite /build/resourceTmp/foo.txt (previously copied from src/main/resources/foo.txt) 
> ```
> 
> For even better clarity track the current task and modifed times so that one could understand
> 
> ```
> src/main/resources/foo.txt (task:mungeResources) attempting to overwrite /build/resourceTmp/foo.txt (previously copied from src/main/resources/foo.txt (task:copyResources) 
> ```
> 
> And possibly NOT fail if the mod time hadn't changed.
> 
> Wait, that's too much memory you say... possibly which is why the biggest problem here is this should be an optional feature for folks who are having issues with things getting mysteriously overwritten. Maybe something like
> 
> ```
> allProjects {
>   strictCopy true
> }
> ```

Thank you but we are discussing a bug in Gradle whereby an error is thrown when a duplicate does not exist.

**Comment by D-Kumar19** [comment]
`...is a duplicate but no duplicate handling strategy has been set. Please refer to https://docs.gradle.org/7.5/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:duplicatesStrategy for details.`

This is the problem that I am receiving when I try to build my project in which I am trying to upgrade the Gradle version from 6.7 to 7.5. I tried to add this in **build.gradle** but the error still persists.

`allprojects {

    apply plugin: 'java'

    tasks.withType(Copy).all {
        duplicatesStrategy 'exclude'
    }
    ...
`

I also tried to add this in my **build.gradle.in** file but I am not able to remove this error and every time build fails. Even tried it with .INCLUDE and .EXCLUDE but no way to get rid of this error. 
`tasks.withType(Copy) {
    duplicatesStrategy = DuplicatesStrategy.WARN
}`

There are two **build.gradle.in** files but the error is from first file. I have tried to add into 1 file in which there is error and also tried it in both files but no chance.

**Comment by spyro2000** [comment]
It's another magic line I will handle to my simple Spring Boot project to make it work with Kotlin and QueryDSL. Don't really know, what causes the actual problem but I am just tired of this.

**Comment by carstenSpraener** [comment]
is it possible this bug is still alive? Having some issues with gradle 8.1.1
see https://github.com/gradle/gradle/issues/25208

Any help is very welcome cause i can't publish at the moment. This is a show stopper

**Comment by GregMarshBYKA** [comment]
I don't know why you can't publish here carstenSpraener (I can't reference you either) - this itself is an issue but to your point, I have not seen anything published that refers to this being resolved and everything so far that I have seen is a workaround for a "NON ERROR". 
I am hoping that something posted here or in the releases will specifically address this issue for we are still using the old version of Gradle.

**Comment by poo0054** [comment]
> 案是：`log4`

I found that you are too right, I have been using the maven, but spring in the use of gradle. I would like to learn a little, I found that it is too dissuasive, no matter what, I never clone the code down after the direct start success, each time there will be a lot of small problems, and finally I really do not want to touch the

**Comment by Azahe** [comment]
I am not sure if anyone of the gradle team watches this task anymore - however maybe this can useful for others, our solution:  

```groovy
subprojects { p ->
  // To prevent naming conflicts in multi-module projects, include the full relative path in the jar name.
  // This is crucial when common module names (like 'api' or 'core') exist in various subdirectories.
  // This strategy ensures jar name uniqueness across the project.
  tasks.withType(Jar).tap {
    configureEach {
      archiveBaseName.set(project.path.replace(':', '-').stripMargin('-'))
    }
  }
}
```

In a situation where two sub modules have same name (as described in the comment) default value for jar files causes error in tasks where these are collected together (for example Spring's `bootWar` task which I assume many people relay on in one way or another).  

This is direct result of `Project#getName` method returning potentially non unique value for multiproject file being used as a default for archive base name.
One potential solution would be to change that default to `Project#getPath` - which according to documentation returns unique value in the repository.  

In this case meddling with `duplicatesStrategy` seems misguided - as allowing duplicates can lead to unintended side effects - like duplicates that developers don't know about.  

This doesn't address conflicts from "outside-of-repo" jars but still looks like an overall improvement.

**Comment by GregMarshBYKA** [comment]
> I am not sure if anyone of the gradle team watches this task anymore - however maybe this can useful for others, our solution (in root `build.gradle` file):
> 
> ```groovy
> subprojects {
>   // To prevent naming conflicts in multi-module projects, include the full relative path in the jar name.
>   // This is crucial when common module names (like 'api' or 'core') exist in various subdirectories.
>   // This strategy ensures jar name uniqueness across the project.
>   tasks.withType(Jar).tap {
>     configureEach {
>       archiveBaseName.set(project.path.replace(':', '-').stripMargin('-'))
>     }
>   }
> }
> ```
> 
> In a situation where two sub modules have same name (as described in the comment) default value for jar files causes error in tasks where these are collected together (for example Spring's `bootWar` task which I assume many people relay on in one way or another).
> 
> This is direct result of `Project#getName` method returning potentially non unique value for multiproject file being used as a default for archive base name. One potential solution would be to change that default to `Project#getPath` - which according to documentation returns unique value in the repository.
> 
> In this case meddling with `duplicatesStrategy` seems misguided - as allowing duplicates can lead to unintended side effects - like duplicates that developers don't know about.
> 
> This doesn't address conflicts from "outside-of-repo" jars but still looks like an overall improvement. Also changing `:` to `-` can maybe be skipped.

@Azahe , thank you for your code sample which addresses issues that I am sure people will benefit from. Unfortunately, the core issue here is not related - Gradle 7 broke something and flags duplicates when there aren't any. This is the real issue and as per your comment at the beginning, I do not know if anyone is looking at this for there does not seem to be any further input. I hope this topic stays open and is not closed due to lack of responses for as far as I can tell nothing has changed.

**Comment by lkoe** [comment]
I'm the orignal bug author and reported a very specific situation, where adding the same source directory twice to the project (something that multiple third-party plugins might do) leads to erroneous reporting of duplicates, as demonstrated by a reproducer project attached.
This specific misbehaviour/bug has been closed and flagged as no-fix by the gradle devs (https://github.com/gradle/gradle/issues/17236#issuecomment-856806005). So there probably won't be happening anything regarding the original issue.

For some reason this specific report seems to be found by people dealing with all sorts of duplicate issues and has become sort of a fly trap since.
I can only recommend to file a separate issue for *your* specific problem, ideally with a reproducer sample attached.

**Comment by GregMarshBYKA** [comment]
@lkoe Hi Lars, there are many variations added but the core issue that you pointed out resulting from the requirement to add a "duplicatesStrategy" even for nonexistent duplicates is still there and a lot of the comments relate so hopefully someone is taking note. Cheers

**Comment by ov7a** [comment]
Thanks, everyone, for the discussion and sorry for the late reply. 

As @donat pointed out in [his message](https://github.com/gradle/gradle/issues/17236#issuecomment-856806005), the idiomatic way to resolve this is to change Gradle plugins and script to avoid intersecting directories. The default fail-fast strategy is correct and, as some of you indicated, helpful in most cases, so it's unlikely that this behavior will be changed.

However, the error message should definitely be improved to indicate the sources of the duplicates to ease the debugging and fixing the issue.

Another option might be improving the duplicates strategy to ignore exact duplicates (e.g. with matching fingerprints) - this should solve the issue for newbie users. However, it's up to the team responsible for that area to decide.

***

The issue is in the backlog of the relevant team, but this area of Gradle is currently not a focus one, so it might take a while before a fix is made.

**Comment by Annonymus-v02** [comment]
I agree that the error message needs to be greatly improved. I have a project which keeps getting spurious duplicate errors on files that don't exist in my project. running `clean` usually but not always fixes it for the next build, but sooner or later it returns and it affects the CI too. The error message specifies the path of the supposedly duplicate file, file the file (which is different every time — the example I'm looking at right now is `META-INF/versions/9/module-info.class`) never actually exists in my project. Presumably it is generated by gradle itself or a plugin, but there is no information whatsoever in the error message with which to debug the cause of the duplicate or even where to find the file in question (with the hope that perhaps looking at the file itself can offer some insight). I'm forced to set duplicatesStrategy=WARN now because I can't eliminate the cause and the CI build risks spuriously failing otherwise. Forcing us to disable (effective) duplicate checking defeats the point of checking for duplicates in the first place.
